### PR TITLE
Add support for v1alpha2 route rules

### PR DESCRIPTION
**Pull request by ijsnellf**
**What this PR does / why we need it**: This PR brings v1alpha2 route rule support to parity with v1alpha1 route rules. Not all v1alpha2 features are implemented. Gateways, named ports, and TCP routes will be implemented in a future PR. Integration tests still use v1alpha1 route rules as well.

**Which issue this PR fixes**: contributes to https://github.com/istio/istio/issues/1925

**Special notes for your reviewer**: 

**Release note**:

```release-note
Added support for v1alpha2 route rules.
```

### CODE REVIEW

File: pilot/adapter/config/crd/generate.sh [hunk 1]
```diff
@@ -47,7 +47,7 @@ var knownTypes = map[string]struct {
 }{
 EOF
 
-CRDS="MockConfig RouteRule IngressRule EgressRule DestinationPolicy HTTPAPISpec HTTPAPISpecBinding QuotaSpec QuotaSpecBinding"
+CRDS="MockConfig RouteRule V1alpha2RouteRule IngressRule EgressRule DestinationPolicy HTTPAPISpec HTTPAPISpecBinding QuotaSpec QuotaSpecBinding"
```
Comment (rshriram) [inline pilot/adapter/config/crd/generate.sh]: I don't know much about the CRD internals to comment. This sounds strange though. I thought the idea behind doing v1alpha2 was to be able to have same CRD, but allow people to use different API versions?

File: pilot/model/config.go [hunk 2]
```diff
@@ -386,6 +398,23 @@ func ResolveHostname(meta ConfigMeta, svc *routing.IstioService) string {
...
+// TODO: support IP addresses (?)
+// TODO: support non-FQDN names in Consul/Eureka
+// support 3 forms
+// reviews => reviews.default.svc.cluster.local
+// reviews.sales => reviews.sales.svc.cluster.local
+// reviews.default.svc.cluster.local => reviews.default.svc.cluster.local (no change)
+func ResolveFQDN(meta ConfigMeta, name string) string {
+	switch strings.Count(name, ".") {
+	case 0:
+		return fmt.Sprintf("%s.%s.svc.%s", name, meta.Namespace, meta.Domain)
```
Comment (rshriram) [inline pilot/model/config.go]: We don't need this precomputation anymore (more like we can't). We should probably have a platform's domain as part of service model or something (e.g., service.consul or svc.cluster.local). You resolve the namespace based on the service (source) where the route rule is applied. So, its destinationname.source-namespace.... or its just destinationname.service.consul ...

File: pilot/model/config.go [hunk 3]
```diff
@@ -433,6 +463,12 @@ func SortRouteRules(rules []Config) {
 }
 
 func (store *istioConfigStore) RouteRules(instances []*ServiceInstance, destination string) []Config {
```
Comment (rshriram) [inline pilot/model/config.go]: We can rename this as RouteRulesV1Alpha1.. in fact, along the lines of @ZackButcher 's PR, I would do another PR that renamed all these functions to v1alpha1routerules.

Or, we could have a simple if-else inside the RouteRules. Based on some static settings in pilot (v1/v2 api), we pick the right variant. This way, function names remain constant?

File: pilot/model/config.go [hunk 4]
```diff
@@ -459,7 +495,57 @@ func (store *istioConfigStore) RouteRules(instances []*ServiceInstance, destinat
...
+			continue
+		}
+
+		var subset bool
+		SourceLoop:
+		for _, endpoint := range rule.Source {
+			if endpoint != nil {
+				for _, instance := range instances {
+					if Labels(endpoint.Labels).SubsetOf(instance.Labels) {
+						subset = true
```
Comment (rshriram) [inline pilot/model/config.go]: going to be slow...

File: pilot/model/config.go [hunk 5]
```diff
@@ -459,7 +495,57 @@ func (store *istioConfigStore) RouteRules(instances []*ServiceInstance, destinat
...
+		out = append(out, config)
+	}
+
+	return out
+}
+
 func (store *istioConfigStore) RouteRulesByDestination(instances []*ServiceInstance) []Config {
+	configs := store.routeRulesByDestinationV1Alpha1(instances)
+	configs = append(configs, store.routeRulesByDestinationV1Alpha2(instances)...)
+	return configs
```
Comment (rshriram) [inline pilot/model/config.go]: ? why process both?

File: pilot/adapter/config/crd/generate.sh (hunk 1 above)
Comment (ijsnellf) [reply pilot/adapter/config/crd/generate.sh]: ah, yes. that would make more sense.

File: pilot/model/BUILD [hunk 6]
```diff
@@ -27,6 +26,8 @@ go_library(
         "@io_istio_api//mixer/v1:go_default_library",
         "@io_istio_api//mixer/v1/config/client:go_default_library",
         "@io_istio_api//routing/v1alpha1:go_default_library",
+        "@io_istio_api//routing/v1alpha2:go_default_library",
+        "//pilot/model/test:go_default_library",
```
Comment (nmittler) [inline pilot/model/BUILD]: Is this the correct ordering? Do we have a lint tool for BUILD files to automatically correct this?

Comment (ijsnellf) [reply pilot/model/BUILD]: I don't know. I think Intellij automatically re-ordered them, but I don't know if they were in any sort of ordering before.

File: pilot/model/config_test.go [hunk 7]
```diff
@@ -263,6 +264,36 @@ func TestResolveHostname(t *testing.T) {
...
+		},
+		{
+			name: "hello.default.svc.cluster.local",
+			meta: model.ConfigMeta{Namespace: "default", Domain: "cluster.local"},
+			fqdn: "hello.default.svc.cluster.local",
+		},
+	}
+
+	for _, test := range cases {
+		if got := model.ResolveFQDN(test.meta, test.name); got != test.fqdn {
```
Comment (ZackButcher) [inline pilot/model/config_test.go]: Here and in the other table driven tests in the PR, please wrap the test body in a `t.Run(func(t *testing.T){ ... })`. [It makes the test harder to screw up by e.g. using `t.Fatal` in the future.](https://blog.golang.org/subtests)

File: pilot/model/BUILD [hunk 8]
```diff
@@ -61,12 +62,12 @@ go_test(
...
     ],
     deps = [
-        ":go_default_library",
-        "//pilot/adapter/config/memory:go_default_library",
-        "//pilot/test/mock:go_default_library",
         "@com_github_davecgh_go_spew//spew:go_default_library",
         "@io_istio_api//mixer/v1:go_default_library",
         "@io_istio_api//mixer/v1/config/client:go_default_library",
         "@io_istio_api//routing/v1alpha1:go_default_library",
+        ":go_default_library",
```
[the remark this answers is not in the document]
Comment (ijsnellf) [reply pilot/model/BUILD]: Thanks. I was not sure what the canonical ordering was supposed to be. Apparently my Intellij Bazel plugin automatically reorders the imports based on different criteria.

File: pilot/model/config.go [hunk 9]
```diff
@@ -21,8 +21,11 @@ import (
 
 	"github.com/golang/protobuf/proto"
 
+	"strings"
```
[the remark this answers is not in the document]
Comment (ijsnellf) [reply pilot/model/config.go]: I'll make sure I pass all the linters etc before I remove the WIP

File: pilot/model/config.go (hunk 2 above)
Comment (ijsnellf) [reply pilot/model/config.go]: Could you elaborate on this? I am not sure what you mean.

Comment (rshriram) [reply pilot/model/config.go]: Can we plumb the platform type into the ResolveFQDN function? So that if its Kube, we can use Namespace and svc.cluster.local from the rule's metadata field.. if its Consul, we use service.consul. If its Eureka, don't do anything. If its CloudFoundry, dont do anything..

Comment (rshriram) [reply pilot/model/config.go]: Because, I would like to use a simple service name in Consul environment as well.

Comment (rshriram) [reply pilot/model/config.go]: Alternatively, you could tweak this logic to check if namespace is emtpy.. and if so, do name+domain.. if its not emtpy, do, name+namespace+domain

File: pilot/model/config.go [hunk 10]
```diff
@@ -457,7 +486,8 @@ func MatchSource(meta ConfigMeta, source *routing.IstioService, instances []*Ser
 	return false
 }
 
-// SortRouteRules sorts a slice of rules by precedence in a stable manner
+// SortRouteRules sorts a slice of v1alpha1 rules by precedence in a stable manner.
+// non-v1alpha1 rules are sorted low without a guaranteed relative ordering.
```
Comment (rshriram) [inline pilot/model/config.go]: ? Its an array right?

File: pilot/model/config.go [hunk 11]
```diff
@@ -471,6 +501,12 @@ func SortRouteRules(rules []Config) {
 }
 
 func (store *istioConfigStore) RouteRules(instances []*ServiceInstance, destination string) []Config {
+	configs := store.routeRules(instances, destination)
+	configs = append(configs, store.routeRulesV2(instances, destination)...)
+	return configs
+}
+
+func (store *istioConfigStore) routeRules(instances []*ServiceInstance, destination string) []Config {
```
Comment (rshriram) [inline pilot/model/config.go]: Add a TODO here ..
//TODO: remove once transition is complete.

We are not going to support both v1alpha1 and v1alpha2 simultaneously.

File: pilot/model/config.go [hunk 12]
```diff
@@ -497,7 +533,66 @@ func (store *istioConfigStore) RouteRules(instances []*ServiceInstance, destinat
...
+				break
+			}
+		}
+		if !found {
+			continue
+		}
+
+		if !matchSource(rule, instances) {
+			continue
+		}
```
Comment (rshriram) [inline pilot/model/config.go]: Not sure if this applies here anymore.. We have an array of rules. So we probably have to do the source match when iterating over individual rules.

File: pilot/model/config.go [hunk 13]
```diff
@@ -497,7 +533,66 @@ func (store *istioConfigStore) RouteRules(instances []*ServiceInstance, destinat
...
+
+		out = append(out, config)
+	}
+
+	return out
+}
+
+// TODO: can the instance matching be optimized?
+func matchSource(rule *routingv2.RouteRule, instances []*ServiceInstance) bool {
```
Comment (rshriram) [inline pilot/model/config.go]: Yeah, we definitely need to do this on a per rule basis. Secondly, since we have settled on a sidecar per pod model, we can safely get rid of the array of instances and the associated for loop. That said, lets do the matchSource when applying individual rules in the array.

File: pilot/model/config_test.go [hunk 14]
```diff
@@ -263,6 +266,38 @@ func TestResolveHostname(t *testing.T) {
...
+		fqdn string
+	}{
+		{
+			name: "hello",
+			meta: model.ConfigMeta{Namespace: "default", Domain: "cluster.local"},
+			fqdn: "hello.default.svc.cluster.local",
+		},
+		{
+			name: "hello.default",
+			meta: model.ConfigMeta{Namespace: "default", Domain: "cluster.local"},
```
Comment (rshriram) [inline pilot/model/config_test.go]: Use Namespace: foobar here. What do we expect? The result should be hello.default.svc.cluster.local.

File: pilot/model/config_test.go [hunk 15]
```diff
@@ -263,6 +266,38 @@ func TestResolveHostname(t *testing.T) {
...
+			fqdn: "hello.default.svc.cluster.local",
+		},
+		{
+			name: "hello.default.svc.cluster.local",
+			meta: model.ConfigMeta{Namespace: "default", Domain: "cluster.local"},
```
Comment (rshriram) [inline pilot/model/config_test.go]: use different namespace here.

File: pilot/model/config_test.go [hunk 16]
```diff
@@ -263,6 +266,38 @@ func TestResolveHostname(t *testing.T) {
...
+			fqdn: "hello.default.svc.cluster.local",
+		},
```
Comment (rshriram) [inline pilot/model/config_test.go]: When you implement the if namespace, name+namespace+domain, else name+domain logic, add test cases appropriately.

File: pilot/model/validation.go [hunk 17]
```diff
@@ -1624,6 +1622,217 @@ func ValidateEndUserAuthenticationPolicySpecBinding(msg proto.Message) error {
...
+			errs = appendErrors(errs, ValidateDuration(http.Timeout))
+		}
+	}
+
+	// TODO: validate once implemented
+	if len(routeRule.Tcp) > 0 {
+		errs = appendErrors(errs, errors.New("TCP route rules have not been implemented"))
+	}
+
+	return
```
Comment (rshriram) [inline pilot/model/validation.go]: May be refactor into two functions like validateHTTPRoute, and validateTCPRoute.. and then call them from this function, per iteration in the for loop ?

File: pilot/proxy/envoy/config.go [hunk 18]
```diff
@@ -542,10 +543,13 @@ func buildDestinationHTTPRoutes(service *model.Service,
...
 		rules := config.RouteRules(instances, service.Hostname)
 		// sort for output uniqueness
+		// if v1alpha2 rules are returned, len(rules) <= 1 is guaranteed
+		// because v1alpha2 rules are unique per host.
 		model.SortRouteRules(rules)
+
 		for _, rule := range rules {
-			httpRoute := buildHTTPRoute(rule, service, servicePort)
-			routes = append(routes, httpRoute)
+			httpRoutes := buildHTTPRoute(rule, service, servicePort)
```
Comment (rshriram) [inline pilot/proxy/envoy/config.go]: pass instances here. for source based routing

File: pilot/proxy/envoy/config.go [hunk 19]
```diff
@@ -542,10 +543,13 @@ func buildDestinationHTTPRoutes(service *model.Service,
 		useDefaultRoute := true
 		rules := config.RouteRules(instances, service.Hostname)
 		// sort for output uniqueness
+		// if v1alpha2 rules are returned, len(rules) <= 1 is guaranteed
+		// because v1alpha2 rules are unique per host.
 		model.SortRouteRules(rules)
+
```
Comment (rshriram) [inline pilot/proxy/envoy/config.go]: Add a TODO to eliminate the for loop below since we now have only a single rule per destination.. Alternatively, we can fold the buildHTTPRoute logic under this function..

File: pilot/proxy/envoy/header.go [hunk 20]
```diff
@@ -58,6 +59,59 @@ func buildHTTPRouteMatch(matches *routing.MatchCondition) *HTTPRoute {
...
 }
 
+func buildHTTPRouteMatches(matches []*routingv2.HTTPMatchRequest) []*HTTPRoute {
+	if len(matches) == 0 {
+		return []*HTTPRoute{{Prefix: "/"}}
+	}
+
+	routes := make([]*HTTPRoute, 0, len(matches))
+	for _, match := range matches {
+		route := &HTTPRoute{}
```
Comment (rshriram) [inline pilot/proxy/envoy/header.go]: I think this can be refactored in a cleaner way. iterate through list of routes, and for each route, iterate through list of matches, and for each match, run buildHttpRoute()..

File: pilot/model/config.go (hunk 10 above)
Comment (ijsnellf) [reply pilot/model/config.go]: Most of the time in Go you are technically using [slices](https://blog.golang.org/go-slices-usage-and-internals), not arrays.

From the Go blog:
> Slices are analogous to arrays in other languages, but have some unusual properties.

File: pilot/model/config.go (hunk 11 above)
Comment (ijsnellf) [reply pilot/model/config.go]: I opted not to put any of the TODOs for the transition into this PR because I thought it would be too verbose. I can add them, but I think almost every function I added will need a TODO.

File: pilot/model/config_test.go [hunk 21]
```diff
@@ -357,61 +391,89 @@ func (errorStore) Delete(typ, name, namespace string) error {
...
+				t.Error(err)
+			}
+			if out := store.RouteRules([]*model.ServiceInstance{instance}, mock.WorldService.Hostname, mock.HelloProxyV0.Domain); len(out) != 1 ||
+				!reflect.DeepEqual(tc.spec, out[0].Spec) {
+				t.Errorf("RouteRules() => expected %#v but got %#v", tc.spec, out)
+			}
+			if out := store.RouteRules([]*model.ServiceInstance{instance}, mock.HelloService.Hostname, mock.HelloProxyV0.Domain); len(out) != 0 {
+				t.Error("RouteRules() => expected no match for destination-matched rules")
+			}
+			// TODO how to handle domain field here
```
Comment (rshriram) [inline pilot/model/config_test.go]: more explanation?

File: pilot/proxy/envoy/route.go [hunk 22]
```diff
@@ -246,6 +282,158 @@ func buildHTTPRoute(config model.Config, service *model.Service, port *model.Por
...
+
+			if len(http.Route) == 0 { // build default cluster
+				cluster := buildOutboundCluster(defaultDestination, port, nil)
+				route.Cluster = cluster.Name
+				route.clusters = append(route.clusters, cluster)
+			} else {
+				route.WeightedClusters = &WeightedCluster{Clusters: make([]*WeightedClusterEntry, 0, len(http.Route))}
+				for _, dst := range http.Route {
+					fqdn := model.ResolveFQDN(dst.Destination.Name, domain)
+					cluster := buildOutboundCluster(fqdn, port, dst.Destination.Labels) // TODO: support Destination.Port
```
Comment (rshriram) [inline pilot/proxy/envoy/route.go]: This is a big todo.. is it the cae that we are not supplying port information in the registration key?

File: pilot/proxy/envoy/route.go [hunk 23]
```diff
@@ -246,6 +282,158 @@ func buildHTTPRoute(config model.Config, service *model.Service, port *model.Por
...
+				protoDurationToMS(http.Timeout) > 0 {
+				route.TimeoutMS = protoDurationToMS(http.Timeout)
+			}
+
+			route.RetryPolicy = buildRetryPolicy(http.Retries)
+
+			applyRedirect(route, http.Redirect)
+			applyRewrite(route, http.Rewrite)
+
+			// Add the fault filters, one per cluster defined in weighted cluster or cluster
```
Comment (rshriram) [inline pilot/proxy/envoy/route.go]: Add a todo here to make fault filters constrained to a specific cluster

File: pilot/proxy/envoy/testdata/fault-route-v1alpha2.yaml.golden [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+  - match:
+      - headers:
+          scooby:
+            exact: doo
+          animal:
+            prefix: dog.cat
+          name:
+            regex: "sco+do+"
+        sourceLabels:
+            version: v0
```
Comment (rshriram) [inline pilot/proxy/envoy/testdata/fault-route-v1alpha2.yaml.golden]: can you add a real multi match? I just want to see multiple routes being generated.

File: pilot/model/validation.go [hunk 25]
```diff
@@ -1624,6 +1622,214 @@ func ValidateEndUserAuthenticationPolicySpecBinding(msg proto.Message) error {
...
+	}
+
+	errs = appendErrors(errs, ValidatePercent(abort.Percent))
+
+	switch abort.ErrorType.(type) {
+	case *routingv2.HTTPFaultInjection_Abort_GrpcStatus:
+		// TODO: GRPC status validation
+		errs = multierror.Append(errs, fmt.Errorf("gRPC fault injection not supported yet"))
+	case *routingv2.HTTPFaultInjection_Abort_Http2Error:
+		// TODO: HTTP2 error validation
```
[the remark this answers is not in the document]
Comment (ijsnellf) [reply pilot/model/validation.go]: done

File: pilot/model/validation.go (hunk 17 above)
Comment (ijsnellf) [reply pilot/model/validation.go]: done

File: pilot/model/validation_test.go [hunk 26]
```diff
@@ -2192,3 +2192,212 @@ func TestValidateTlsOptions(t *testing.T) {
...
+			AllowHeaders:  []string{"header1", "header2"},
+			ExposeHeaders: []string{"header3"},
+			MaxAge:        &duration.Duration{Seconds: 2},
+		}, valid: false},
+		{name: "bad header", in: &routingv2.CorsPolicy{
+			AllowMethods:  []string{"GET", "POST"},
+			AllowHeaders:  []string{"header1", "header2"},
+			ExposeHeaders: []string{"HEADER3"},
+			MaxAge:        &duration.Duration{Seconds: 2},
+		}, valid: false},
```
Comment (ZackButcher) [inline pilot/model/validation_test.go]: Add a test for invalid max age please.

File: pilot/model/validation_test.go [hunk 27]
```diff
@@ -2192,3 +2192,212 @@ func TestValidateTlsOptions(t *testing.T) {
...
+			HttpDelayType: &routingv2.HTTPFaultInjection_Delay_FixedDelay{
+				FixedDelay: &duration.Duration{Seconds: 3},
+			},
+		}, valid: true},
+		{name: "invalid percent", in: &routingv2.HTTPFaultInjection_Delay{
+			Percent: 101,
+			HttpDelayType: &routingv2.HTTPFaultInjection_Delay_FixedDelay{
+				FixedDelay: &duration.Duration{Seconds: 3},
+			},
+		}, valid: false},
```
Comment (ZackButcher) [inline pilot/model/validation_test.go]: Add cases for:
- invalid delay
- nil Abort and nil Delay

File: pilot/proxy/envoy/route.go (hunk 22 above)
Comment (ijsnellf) [reply pilot/proxy/envoy/route.go]: I'll create an issue to track this

File: pilot/proxy/envoy/testdata/fault-route-v1alpha2.yaml.golden (hunk 24 above)
Comment (ijsnellf) [reply pilot/proxy/envoy/testdata/fault-route-v1alpha2.yaml.golden]: added it to `multi-match-fault-v1alpha2.yaml.golden`
